### PR TITLE
Signing and validation of pandora sharding chain header hash and extraData

### PR DESCRIPTION
**Pull request by atif-konasl**
### Related Issue No- [#2](https://github.com/lukso-network/vanguard-consensus-engine/issues/2) 

This PR implements header hash and extra field validation of pandora sharding chain when validator will propose a new consensus block. Currently pandora sharding info does not store into consensus block. It just validates the pandora sharding chain header info as well as signs on header hash and send it back to pandora sharding chain.

When `vanguard` run with `validator` and otherside `pandora` also need to run to get effect.

By default, `validator` client communicates through `http` provider in `http://127.0.0.1:8545` endpoint.

### CODE REVIEW

File: validator/client/propose.go [hunk 1]
```diff
@@ -81,6 +97,15 @@ func (v *validator) ProposeBlock(ctx context.Context, slot types.Slot, pubKey [4
 		return
 	}
 
+	// processPandoraShardHeader method process the block header from pandora chain
+	if status, err := v.processPandoraShardHeader(ctx, b, slot, epoch, pubKey); !status || err != nil {
```
Comment (frozeman) [inline validator/client/propose.go:101]: I am not 100% sure if we should use the word `pandora` in our function names and not simple: `v.processShardHeader`

as in the future other systems could create them too. Whats your ideas here @blazejkrzak @mxmar ?
Maybe it makes sense to keep it pandora specific, as for other shard types we will have a new name...

File: validator/client/propose.go [hunk 2]
```diff
@@ -330,3 +355,81 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [48]byte) ([]byte, e
...
+		return false, err
+	}
+	header.MixDigest = common.BytesToHash(headerHashSig)
+	var headerHashSig32Bytes [32]byte
+	copy(headerHashSig32Bytes[:], headerHashSig)
+	return v.pandoraService.SubmitShardBlockHeader(ctx, header.Nonce.Uint64(), headerHash, headerHashSig32Bytes)
+}
+
+// verifyPandoraHeader verifies header hash and extraData field
+func (v *validator) verifyPandoraHeader(beaconBlk *ethpb.BeaconBlock, slot types.Slot, epoch types.Epoch,
```
Comment (frozeman) [inline validator/client/propose.go]: either we hard `...ShardHeader` everywhere, or nowhere. :) see function above
Lets better call it `verifyPandoraShardHeader`

**Review by frozeman — APPROVED** [review]
Lets make these changes and we are good.

File: validator/client/propose.go (hunk 1 above)
Comment (blazejkrzak) [reply validator/client/propose.go:101]: `processShardHeader`, then if another being comes around use strategy pattern

File: validator/client/propose.go (hunk 2 above)
Comment (blazejkrzak) [reply validator/client/propose.go]: private functions sometimes are used as separation of certain piece of code, in my opinion as log as it is descriptive i'm fine.
Altough `shardHeader` could be nice convention to stay with

File: validator/keymanager/types.go [hunk 3]
```diff
@@ -19,6 +19,8 @@ type IKeymanager interface {
 	Sign(context.Context, *validatorpb.SignRequest) (bls.Signature, error)
 	// SubscribeAccountChanges subscribes to changes made to the underlying keys.
 	SubscribeAccountChanges(pubKeysChan chan [][48]byte) event.Subscription
+	// SignHeaderHash signs pandora chain header hash using a validator key
+	SignHeaderHash(ctx context.Context, req *validatorpb.SignRequest) ([]byte, error)
```
Comment (blazejkrzak) [inline validator/keymanager/types.go]: it would be great to have a test for it since it is interface function

**Review by blazejkrzak — APPROVED** [review]

**Comment by blazejkrzak** [comment]
Looks good to me

File: validator/client/propose.go (hunk 1 above)
Comment (atif-konasl) [reply validator/client/propose.go:101]: My guess, in future other sharding chain will not be verified as like as pandora chain.

File: validator/keymanager/types.go (hunk 3 above)
Comment (atif-konasl) [reply validator/keymanager/types.go]: Okay thanks.

File: validator/client/propose.go [hunk 4]
```diff
@@ -330,3 +355,82 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [48]byte) ([]byte, e
...
+	copy(headerHashSig32Bytes[:], compressedSig)
+	return v.pandoraService.SubmitShardBlockHeader(ctx, header.Nonce.Uint64(), headerHash, headerHashSig32Bytes)
+}
+
+// verifyPandoraShardHeader verifies pandora sharding chain header hash and extraData field
+func (v *validator) verifyPandoraShardHeader(beaconBlk *ethpb.BeaconBlock, slot types.Slot, epoch types.Epoch,
+	header *eth1Types.Header, headerHash common.Hash, extraData *pandora.ExtraData) error {
+
+	// verify header hash
+	if header.Hash() != headerHash {
```
Comment (blazejkrzak) [inline validator/client/propose.go:410]: This is invalid

File: validator/client/propose.go [hunk 5]
```diff
@@ -330,3 +355,82 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [48]byte) ([]byte, e
...
+		log.WithError(errInvalidSlot).Error("invalid slot from pandora chain")
+		return errInvalidSlot
+	}
+	// verify epoch number
+	if extraData.Epoch != uint64(epoch) {
+		log.WithError(errInvalidEpoch).Error("invalid epoch from pandora chain")
+		return errInvalidEpoch
+	}
+	// verify proposer index
+	if extraData.ProposerIndex != uint64(beaconBlk.ProposerIndex) {
```
Comment (blazejkrzak) [inline validator/client/propose.go:430]: remove this and we are fine

**Pull request merged**